### PR TITLE
Bump swiftshader to 84f5eeb6dd9b225f465f93737fa76aad7de355cf

### PR DESCRIPTION
**Pull request by hanhanW**
- Build and push new images to GCR.
- Update Dockerfile in *-swiftshader to install vulkan utils.

### CODE REVIEW

**Comment by hanhanW** [comment]
Do I need to `git add third_party/swiftshader`?

**Comment by hanhanW** [comment]
> Do I need to `git add third_party/swiftshader`?

I need to add this to pass checks, added.

**Comment by antiagainst** [comment]
The SwiftShader submodule is not actually used by IREE. (It's mostly there for convenience so that one can use the script to compile it directly if needed.) The docker images pins down to a specific SwiftShader version: https://github.com/google/iree/blob/main/build_tools/docker/swiftshader/Dockerfile#L20. I think we need to update there to really update the SwiftShader used by CIs. (The submodule should be updated to keep consistency too.)

**Comment by hanhanW** [comment]
> The SwiftShader submodule is not actually used by IREE. (It's mostly there for convenience so that one can use the script to compile it directly if needed.) The docker images pins down to a specific SwiftShader version: https://github.com/google/iree/blob/main/build_tools/docker/swiftshader/Dockerfile#L20. I think we need to update there to really update the SwiftShader used by CIs. (The submodule should be updated to keep consistency too.)

Good to know this! The MobileBert tests doesn't run on CI, so won't hit the issue even the Dockerfile is not updated. Seems that they are not consistent now, let's fix that part as well. Thanks for pointing this out!

**Comment by antiagainst** [comment]
Yup. We need to update Vulkan related stuff all together in the docker images and GCP base image (related: #3550). SwiftShader should be a part of that.

**Comment by ScottTodd** [comment]
Yeah, we should (separate from this) remove the SwiftShader submodule and update our helper script to fetch the source and build on-demand. The submodule use was originally modeled after the Bazel setup I wrote where `--config=vk_swiftshader` would automatically build and configure SwiftShader as we have it internally. We have more flexibility here in open source land to configure our test and development environments outside of the context of a single `bazel` command.

File: build_tools/docker/swiftshader/Dockerfile
Comment (GMNGeoffrey) [inline build_tools/docker/swiftshader/Dockerfile:20]: Please also rebuild the docker images and update the commit used in the RBE image (which has to be separate because it uses Ubuntu 16.04). See https://github.com/google/iree/blob/main/build_tools/docker/README.md I might need to give you some permissions in our cloud project

**Review by GMNGeoffrey — CHANGES_REQUESTED** [review]

File: SUBMODULE_VERSIONS
Comment (ScottTodd) [inline SUBMODULE_VERSIONS:16]: If we're going through the effort to update this, can you bump this again, at least past https://github.com/google/swiftshader/commit/df17a76102dfabb3f1bd6e51449cece9f77b45e3 ? That might let us remove some workarounds like
https://github.com/google/iree/blob/8bda4d9a1241c7716ac6ed0f9738ff57a5e3c25d/iree/hal/cts/cts_test_base.h#L33-L42

**Review by ScottTodd — CHANGES_REQUESTED** [review]

**Comment by GMNGeoffrey** [comment]
FYI Phoenix deleted the swiftshader submodule in https://github.com/google/iree/commit/b214d23b9521e240af634aaef380111942a25f53 and updated the script to match the dockerfiles at https://github.com/google/swiftshader/commit/e0de282e678b72dac25b81f8d25305231aec01f8, which I think at least addresses the mobile bert issues. We might still want to bump it for Scott's other request.

**Comment by hanhanW** [comment]
Just found that I forgot this PR, I can take a look at this later this week. I'll be happy if others want to pick this up. Thanks for the update!

File: SUBMODULE_VERSIONS
Comment (hanhanW) [reply SUBMODULE_VERSIONS]: Yes, this is later then df17a76.

**Comment by hanhanW** [comment]
Followed instructions in https://github.com/google/iree/blob/main/build_tools/docker/README.md and update the PR, PTAL.

**Comment by hanhanW** [comment]
Looks like it can't find `vulkaninfo`, not sure what's wrong. :(

**Comment by hanhanW** [comment]
The old one has /usr/bin/vulkaninfo, but the one I created doesn't have it.

**Comment by hanhanW** [comment]
Cool, everything is working now. The only concern is, what vulkan SDK version would we like to use in SwiftShader env? Now we're using `1.2.154` for `*-swiftshader`, and using `1.2.141` for `*-vulkan`. We can either make them the same or leave them as this way.

**Comment by antiagainst** [comment]
> Cool, everything is working now. The only concern is, what vulkan SDK version would we like to use in SwiftShader env? Now we're using `1.2.154` for `*-swiftshader`, and using `1.2.141` for `*-vulkan`. We can either make them the same or leave them as this way.

We should avoid inconsistencies. I'm working on https://github.com/google/iree/pull/3754 to install the Vulkan SDK via downloading the tarball. Hopefully that can get everything updated, including both the normal docker images and RBE (which stays at Ubuntu 16.04). Recent Vulkan SDK versions from `apt` drops support for Ubuntu 16.04.

**Comment by GMNGeoffrey** [comment]
Lei has updated all the vulkan stuff. If we still want to update swiftshader, can you rebase this?

**Review by GMNGeoffrey — CHANGES_REQUESTED** [review]
Please rebase and update this with the vulkan changes

**Comment by hanhanW** [comment]
Chatted with @antiagainst offline. Will wait until the reverted patch land.

File: build_tools/docker/manage_images.py
Comment (GMNGeoffrey) [inline build_tools/docker/manage_images.py:67]: You did not add this dependency. The swiftshader image does not depend on the vulkan image. The swiftshader image is just the one called "swiftshader" that does the swiftshader install build that is copied by other images.

**Review by GMNGeoffrey — CHANGES_REQUESTED** [review]

File: build_tools/docker/bazel-tensorflow-swiftshader/Dockerfile
Comment (hanhanW) [reply build_tools/docker/bazel-tensorflow-swiftshader/Dockerfile:29]: This is needed for check_vulkan.sh, which calls `vulkaninfo`. And we want to use the binaries in the same vulkan SDK?

Comment (GMNGeoffrey) [reply build_tools/docker/bazel-tensorflow-swiftshader/Dockerfile:29]: This is already part of the image because it inherits from https://github.com/google/iree/blob/main/build_tools/docker/bazel-tensorflow-vulkan/Dockerfile

File: build_tools/docker/manage_images.py
Comment (hanhanW) [reply build_tools/docker/manage_images.py]: done

File: build_tools/docker/bazel-tensorflow-swiftshader/Dockerfile
Comment (hanhanW) [reply build_tools/docker/bazel-tensorflow-swiftshader/Dockerfile]: It actually don't work if I revert the change. I can see `vulkaninfo` in `/opt/vulkan-sdk/1.2.154.0/x86_64/bin`, but the ENV is not set. It does fail on CI.

I thought the ENV would be inherit but seems not, any ideas?

Comment (GMNGeoffrey) [reply build_tools/docker/bazel-tensorflow-swiftshader/Dockerfile]: Hmmmm let me investigate

Comment (GMNGeoffrey) [reply build_tools/docker/bazel-tensorflow-swiftshader/Dockerfile]: Ok I have no idea why this is failing, but I got fed up debugging docker images and pinned everything to digests (https://github.com/google/iree/pull/4047). After that, I was able to trivially update the swiftshader version (https://github.com/google/iree/pull/4059). Sorry for all the churn on this PR, but I think it can now be closed in favor of that one.

**Pull request closed without merging**
